### PR TITLE
Complain about invalid sources again

### PR DESCRIPTION
**Pull request by junaire**
This reverts https://github.com/vgvassilev/clang/commit/a5ee33ae48

This patch previous tried to solve some problems with virtual files, but
when trying to upstream it, it brokes the tests below:

* Modules/module-file-modified.c
* Modules/explicit-build-missing-files.cpp

### CODE REVIEW

**Comment by vgvassilev** [comment]
@phsft-bot huild!

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
@smuzaffar, can you test this PR on whole cmssw?

**Comment by junaire** [comment]
Looks like I need a CERN account to see the test results?

**Comment by vgvassilev** [comment]
@phsft-bot build just on mac1015/python3

@junaire, the failures do not seem caused by this patch but likely infrastructure.

**Comment by junaire** [comment]
The Jenkins build seems to be blocked by the clang-format checks, but I can't find anything not well-formatted in my commit...

**Comment by vgvassilev** [comment]
> The Jenkins build seems to be blocked by the clang-format checks, but I can't find anything not well-formatted in my commit...

These particular builders seem to have a huge backlog…

**Comment by junaire** [comment]
Are the failed tests in Jenkins related to my PR? May I see the failed results?

**Comment by vgvassilev** [comment]
> Are the failed tests in Jenkins related to my PR? May I see the failed results?

These are glitches. I think you can see the logs if you make a cern lightweight account. I think this PR is good to go.

**Review by vgvassilev — APPROVED** [review]
It looks like this change is not needed anymore - many of the codepaths have changed over the years and in particular when we landed the C++ modules support. 

cc: @Axel-Naumann

**Comment by pcanal** [comment]
@vgvassilev Do we actually have any tests that checks that things still work if the files used to produce the pcm (or the pch) are gone and/or not longer accessible? (i.e. the test framework would have to explicitly hides/move/removes some files to achieve this).

**Comment by Axel-Naumann** [comment]
There are two parts to this:
- for this PR to be correctly validated we need to have tested a setup where the build directory is not available anymore at runtime. Does CMS test this, @smuzaffar @vgvassilev ?
- we should have a dedicated test for this in ROOT (as standalone cling doesn't use modules). Do we?

**Comment by vgvassilev** [comment]
> @vgvassilev Do we actually have any tests that checks that things still work if the files used to produce the pcm (or the pch) are gone and/or not longer accessible? (i.e. the test framework would have to explicitly hides/move/removes some files to achieve this).

We seem to have builders that use the PCH (nortcxxmod). Then, we have cmssw that has the hybrid model, forward declarations and modules, including relocating the ROOT build - I doubt we would not catch the failure of this patch. I suspect what saves us here is several optimizations we did over the years where rootcling does not produce a lot of virtual file entries and basically pch/pcm files have less (but not 0) virtual files.

@Axel-Naumann, I think you should know better - you created the patch long time ago ;) My memory is that we were trying to fix something with cms/atlas and PCH?

> * for this PR to be correctly validated we need to have tested a setup where the build directory is not available anymore at runtime. Does CMS test this, @smuzaffar @vgvassilev ?

Oh, yes, they do :)

**Comment by smuzaffar** [comment]
@Axel-Naumann , for CMS tests, build directory is not available at runtime. Most of the times build and runtime jenkins jobs run on different machines. Also at build time cmssw external packages ( including `root`) are built on local disk and deployed on /cvmfs for runtime tests.

**Comment by Axel-Naumann** [comment]
Thanks! I hope we can address the relocated build case in our CI once we get @smuzaffar 's help during the 2nd half of this year!

**Comment by smuzaffar** [comment]
yes header files are available via ROOT_INCLUDE_PATH for CMS.

**Comment by pcanal** [comment]
Alright.  @Axel-Naumann @vgvassilev So CMS is not a good test for this.  Do we have any other test?

**Comment by vgvassilev** [comment]
@pcanal, this patch is not checking existence of header files - see the original description of the PR. This patch was intended to check if a file was modified or not... So cms is a test case of this IMO.

I am not sure what you are after here. IIRC, this patch was developed to cover some failures in both cms and root. None of them exist anymore it seems. Are you proposing to continue keeping this patch? If so, the cost of keeping such technical debt is we will have slow llvm upgrades.

**Comment by pcanal** [comment]
I am all for removing unnecessary code.  I did not infer from the description that this was about "out of date" file rather than missing files.   My concern is that I do not know (and "was developed to cover some failures in both cms and root. " is not specific enough to help me here :) ) that we actually have any actual case in routine ROOT or CMSSW testing to cover this case.  If you can point me to where it is tested (and/or the problem(s) description that were fix by this).   In particular if the test is about a file references by the module becoming newer than the produced module ( unless we specifically added a test for it), I can not see how the automated testing would triggered the (previously?) failing case.

TLDR: I do not know that the testing we had done so far prove that the patch is no longer needed (but I of course could be wrong or mis-informed).

**Comment by vgvassilev** [comment]
I don't think we have an explicit test for this patch. We did some archeology. The commit message is rather poor and dates back to 234fd515ed3893629e1e170e5c497688c81281b9. We could only indirectly guess what was trying to do and the best description of the code that was failing and we tried to fix is these 2 failing tests: `clang/test/Modules/module-file-modified.c` and `clang/test/Modules/explicit-build-missing-files.cpp`. Given what fails in these tests and the age of this change I find it highly unlikely that cmssw won't be able to catch such a setup.

At that point, I don't think we are much more informed than you and we are hoping that dropping this change is reasonable to move forward. The good thing is that this is an atomic commit which is easy to bisect if something somewhere starts failing...

**Comment by pcanal** [comment]
Fair enough.  Thanks.
